Configurable HttpClient

masterT opened the pull request:
References:
- #96
- #91 
- #37

## Description

Allow the _Model_ `httpClient`, `jsonApiBaseUrl`, `jsonApiType` to be accessible without an instance using static accessor.

---

```ts
import { Model, AxiosHttpClient } from 'coloquent'
import axios from 'axios'

const customAxioInstance = axios.create(options);

class Artist extends Model {
  static jsonApiBaseUrl = 'http://www.app.com/api';
  static jsonApiType = 'artists';
  static httpClient = new AxiosHttpClient(customAxioInstance);
}

Artist.effectiveJsonApiBaseUrl
// "http://www.app.com/api"

Artist.effectiveJsonApiType
// "artists"

Artist.effectiveHttpClient
// customAxioInstance

Artist.getJsonApiUrl()
// "http://www.app.com/api/artists"
```

DavidDuwaer commented:
Looks good. Implemented the todo for overriding the endpoint with a Model.endpoint property. Streamlined some stuff here and there. Have not added tests yet. Could you write tests for your use case(s)? I guess after that, it's just updating the readme and it should be ready for release.

masterT commented:
I mark the PR as ready but it's not really ready, I wanted your input for the direction I took to solve the issues.

I will update the PR with the missing tests.I plan to test it with in a real side project too.

masterT commented on src/Model.ts:
```diff
@@ -323,33 +327,55 @@ export abstract class Model
-    public getJsonApiType(): string
-    {
-        return this.jsonApiType;
+    /**
+     * @deprecated Use the static method with the same name instead
+     */
+    public getJsonApiBaseUrl(): string {
+      return (this as Model).constructor.getJsonApiBaseUrl()
+    }
```
I'm not sure I would deprecate this method since `type` is part of a JSON:API resource.

DavidDuwaer replied:
(Assuming you mean the comment on `getJsonApiType()`. In JSON:API this property is available on every instance of every resource. On the JSON:API level this is the way to communicate the type of an instance.

However on the "object model" level to which Coloquent translates resources, also the way of communicating the type is translated, i.e. by actually creating an object in the right type, so the type of the model can be obtained by  `.constructor` (or by doing an `instanceof` check). Since the object model is supposed to be an abstraction from the JSON:API/resource layer, this is the type people should use if they want to work with the types of the models. However if one still needs to get not the type of the object itself, but its corresponding JSON:API type, then one can ask the object type rather than the object, as it will vary by type and not by instance, so e.g. do `(new Teacher()).constructor.jsonApiType` or `Teacher.jsonApiType`. As of now `Teacher.getJsonApiType()` is also a public method and a way to do this, but is longer syntax and exists only so that the library internally does a runtime check on the input, so perhaps it should be made private, to simplify the API surface for the user. I'll leave that for another time.

masterT replied:
My bad haha, I meant `getJsonApiType()` like you said. Ok, I understand, you want to keep the abstraction from JSON:API in the model layer. 👍

masterT commented:
@DavidDuwaer I added some tests.

I also added support for
- `endpoint` with or without heading `"/"`
- `jsonApiBaseUrl` with pr without trailing `"/"`

Let me know if there is anything to change.

masterT commented:
I think the heading/trailing "/" manipulation should be done in the json base URL method.

DavidDuwaer commented:
> I think maybe the heading/trailing "/" manipulation should be done in the JSON base URL method, what do you think?

Yes indeed. Nice to think about the feature though. But for the implementation we should keep an immutable approach as much as possible. Mutating the setting in-place will cause some w-t-f debugging moments of users seeing their settings get a life of their own 😛.

I think this PR is ready for release save for the trailing slash. Perhaps revert that, and put it in a separate PR?

DavidDuwaer commented:
Since the configuration API is changing to be all with static properties, I felt some changes had to be made to make the step to the next major version immediately good.

- If some configuration properties are going to be static now, we might as well make them all static
- `Model.configProperty` (fill in either one of the static config properties on Model) and `Model.getConfigProperty()` was not a clear dichotomy. `Model.someProperty` and `Model.effectiveSomeProperty` conveys at least a bit more information. Also moved to an es6 getter there for a more consistent syntax.
- In some cases we were mutating the configuration property in-place (e.g. Model.httpClient). Now not anymore: `Model.configProperty` and `Model.effectiveConfigProperty` will be the same whenever you call them.

A bit many changes, but if you look at them per-commit they should be pretty easy to follow. Do you want to take a shot at reviewing them?

Anyway since it is anyway many changes I'm eyeing a beta release, i.e. 3.0.0-beta. Let me know what you think of the changes and if they're ok, let's release :).

masterT commented:
> A bit many changes, but if you look at them per-commit they should be pretty easy to follow.

Indeed, very easy to follow, thanks! 

It looks good to me, I made some changes in the _Model_ test suite:
- fix test description to reflect the tested method
- add some basic tests for the trailing `"/"` for `jsonApiBaseUrl`
- add some basic tests for the heading `"/"` for `endpoint`

I think it might be a good idea to release a beta first.

masterT commented:
Maybe the next step before the next major release would be to support custom `endpoint` (#76, #70), this means using `getJsonApiUrl()` in the builder (and maybe somewhere else). I think this can be done in another PR.

DavidDuwaer commented:
Available as 3.0.1-beta
